Mesh_3/include/CGAL/Mesh_3/tet_soup_to_c3t3.h

```diff
@@ -0,0 +1,447 @@
+  tr.tds().set_dimension(3);
+  if(!assign_neighbors<Tr>(tr, incident_cells_map))
+    return false;
+
+  std::cout << "built triangulation : " << std::endl;
+  std::cout << tr.number_of_vertices() << " vertices" << std::endl;
+  std::cout << tr.number_of_cells() << " cells" << std::endl;
+
+  BOOST_FOREACH(Vertex_handle vh, vertex_handle_vector)
+  {
+    vh->set_dimension(3);
+  }
+  if(c3t3_loader_failed)
+  {
+    return true;
+  }
+  else
+    return tr.is_valid(true);
+}
+
+template<class Tr, bool c3t3_loader_failed>
+bool build_triangulation_from_file(std::istream& is,
+                                   Tr& tr)
+{
+  typedef typename Tr::Point                                  Point_3;
+
+  typedef boost::array<int, 3> Facet; // 3 = id
+  typedef boost::array<int, 4> Tet; // 4 = id
+  typedef boost::array<int, 5> Tet_with_ref; // first 4 = id, fifth = reference
+
+  std::size_t finite_exterior_cells_counter = 0;
+  std::vector<Tet_with_ref> finite_cells;
+  std::vector<Tet> infinite_cells;
+  std::vector<Point_3> points;
+  std::map<Facet, int> border_facets;
+
+  // grab the vertices
+  int dim;
+  int nv, nf, ntet, ref;
+  std::string word;
+
+  is >> word >> dim; // MeshVersionFormatted 1
+  is >> word >> dim; // Dimension 3
+
+  CGAL_assertion(dim == 3);
+
+  while(is >> word && word != "End")
+  {
+    if(word == "Vertices")
+    {
+      is >> nv;
+      for(int i=0; i<nv; ++i)
+      {
+        double x,y,z;
+        is >> x >> y >> z>>ref;
+        points.push_back(Point_3(x,y,z));
+      }
+    }
+
+    if(word == "Triangles")
+    {
+      is >> nf;
+      for(int i=0; i<nf; ++i)
+      {
+        int n1, n2, n3, surface_patch_id;
+        is >> n1 >> n2 >> n3 >> surface_patch_id;
+        // no use for boundary facets for now...
+        Facet facet;
+        facet[0] =n1 -1;
+        facet[1] =n2 -1;
+        facet[2] =n3 -1;
+        //find the circular permutation that puts the smallest index in the first place.
+        int n0 = (std::min)((std::min)(facet[0],facet[1]), facet[2]);
+        int k=0;
+        Facet f;
+        do
+        {
+          f[0]=facet[(0+k)%3];
+          f[1]=facet[(1+k)%3];
+          f[2]=facet[(2+k)%3];
+          ++k;
+        }while(f[0] != n0);
+        border_facets.insert(std::make_pair(f, surface_patch_id));
+      }
+    }
+
+    if(word == "Tetrahedra")
+    {
+      is >> ntet;
+      for(int i=0; i<ntet; ++i)
+      {
+        int n0, n1, n2, n3, reference;
+        is >> n0 >> n1 >> n2 >> n3 >> reference;
+        Tet_with_ref t;
+        t[0] = n0 - 1;
+        t[1] = n1 - 1;
+        t[2] = n2 - 1;
+        t[3] = n3 - 1;
+        t[4] = reference;
+        finite_cells.push_back(t);
+      }
+    }
+  }
+
+
+
+  if(!finite_exterior_cells_counter)
+  {
+    // The finite interior cells MUST be the convex hulls of the point
+    std::cerr << "WARNING: no finite exterior cell provided..." << std::endl;
+  }
+
+  CGAL_precondition(!finite_cells.empty());
```

Comment:
replace this precondition by a pop-up message "your c3t3 is empty", and return `false`